src/uct/ib/rc/accel/gga_mlx5.c

```diff
@@ -265,23 +254,55 @@ static uct_md_ops_t uct_mlx5_gga_md_ops = {
     .detect_memory_type = ucs_empty_function_return_unsupported,
 };
 
-ucs_status_t uct_ib_mlx5_gga_md_open(struct ibv_device *ibv_device,
-                                     const uct_ib_md_config_t *md_config,
-                                     struct uct_ib_md **md_p)
+static ucs_status_t
+uct_ib_mlx5_gga_md_open(uct_component_t *component, const char *md_name,
+                        const uct_md_config_t *uct_md_config, uct_md_h *md_p)
 {
-    ucs_status_t status;
+    const uct_ib_md_config_t *md_config = ucs_derived_of(uct_md_config,
+                                                         uct_ib_md_config_t);
+    ucs_status_t status = UCS_ERR_UNSUPPORTED;
+    struct ibv_device **ib_device_list, *ib_device;
+    int num_devices, fork_init;
+    uct_ib_md_t *md;
+
+    ucs_trace("opening GGA device %s", md_name);
 
     if (md_config->devx == UCS_NO) {
         return UCS_ERR_UNSUPPORTED;
     }
 
-    status = uct_ib_mlx5_devx_md_open(ibv_device, md_config, md_p);
+    /* Get device list from driver */
+    ib_device_list = ibv_get_device_list(&num_devices);
+    if (ib_device_list == NULL) {
+        ucs_debug("Failed to get GGA device list, assuming no devices are present");
+        status = UCS_ERR_NO_DEVICE;
+        goto out;
+    }
+
+    status = uct_ib_get_device_by_name(ib_device_list, num_devices, md_name,
```

Comment:
maybe unite ibv_get_device_list and uct_ib_get_device_by_name to a single function?

Reply:
I didn't implement this way because the device is a pointer to device list element which is used later in `uct_ib_mlx5_devx_md_open` and similar in IB, so the list cannot be freed in the united function.

Reply:
it's a pointer to a pointer, so should be ok

Reply:
https://linux.die.net/man/3/ibv_get_device_list
```
Client code should open all the devices it intends to use with ibv_open_device() before calling ibv_free_device_list(). Once it frees the array with ibv_free_device_list(), it will be able to use only the open devices; pointers to unopened devices will no longer be valid.
```